Update documentation for agents (#29937)
huggingface/transformers: docs/source/en/agents.md

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->
# Agents and tools

> Large Language Models (LLMs) trained to perform [causal language modeling](./tasks/language_modeling.) can tackle a wide range of tasks, but they often struggle with basic tasks like logic, calculation, and search. The worst scenario is when they perform poorly in a domain, such as math, yet still attempt to handle all the calculations themselves.

Reviewer: It's a bit weird to talk about "the worst scenario" without first having introduced the ability to use tools no?

Here's a reword, lmk what you think: 
```suggestion
Large Language Models (LLMs) trained to perform [causal language modeling](./tasks/language_modeling.) can tackle a wide range of tasks, but they often struggle with basic tasks like logic, calculation, and search. When prompted in domains in which they do not perform well, they often fail to generate the answer we expect them to.
```

Author: Thank you, I adopted this wording instead!

- it takes a `str` as input and returns a `str`
- it will stop generating output _before_ the sequences passed in argument `stop`

You also need a `tools` argument, as a list of `Tools`. We can start with an empty one but specify `add_base_tools` to start from a default set of arguments.

Then you can define your agent and run it.

```python
from transformers import CodeAgent

agent = CodeAgent(llm_engine=llm_engine, tools=[], add_base_tools=True)

agent.run("Draw me a picture of rivers and lakes")
```

The prompt and output parser were automatically defined here, but you can easily inspect them.

![Untitled](Doc%20agents%20ac753b9a3b934eaba22f659ba994c4bd/Untitled%201.png)

```python
agent.prompt_template
```

Note that your agent is powered by a LLM, so small variations in your prompt might yield completely different results. It's important to explain as clearly as possible the task you want to perform. We go more in-depth on how to write good prompts [here](custom_tools#writing-good-user-inputs).

Every [`~Agent.run`] operation is independent, so you can run it several times in a row with different tasks.

### Implementation of agents

#### ReactAgent

> This is the go-to agent to solve reasoning tasks, since the ReAct framework makes it really efficient to think on the basis of its previous observations.

Reviewer: Would be good to share the paper in which it was introduced as well

Author: Good idea! It's added now.

|------------------------------------------------------------------------------------------------------------------------------------------------------|
| <audio controls><source src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/damo.wav" type="audio/wav"/> |

<Tip>

The agent's prompt will reflect the new tool description. But beware when adding tools to an agent that already works well!
It can result in your tool being selected way more than others or for other
tools to be selected instead of the one you have defined.

</Tip>

Replacing existing tools can be done simply by updating the agent's toolbox using method `agent.toolbox.update_tool()`.

<Tip>

Overwriting existing tools can be beneficial if we want to use a custom tool exactly for the same task as an existing tool 
because the agent is well-versed in using the specific task. Beware that the custom tool should follow the exact same API 
as the overwritten tool in this case, or you should adapt the prompt template to make sure all examples using that
tool are updated.

</Tip>

> ### Leverage gradio-tools

Author: Thank you! What's the rationale for this change? And also replacing "Import tools from LangChain" with "LangChain tools"?

[gradio-tools](https://github.com/freddyaboulton/gradio-tools) is a powerful library that allows using Hugging
Face Spaces as tools. It supports many existing Spaces as well as custom Spaces to be designed with it.

> We offer support for `gradio_tools` by using the `Tool.from_gradio` method. For example, we want to take
> advantage of the `StableDiffusionPromptGeneratorTool` tool offered in the `gradio-tools` toolkit so as to
> improve our prompts and generate better images.
> 
> We first import the tool from `gradio_tools` and instantiate it:

Reviewer 2: ```suggestion
Transformers supports `gradio_tools` with the [`Tool.from_gradio`] method. For example, let's use the [`StableDiffusionPromptGeneratorTool`](https://github.com/freddyaboulton/gradio-tools/blob/main/gradio_tools/tools/prompt_generator.py) from `gradio-tools` toolkit for improving prompts to generate better images.

Import and instantiate the tool:
```

```python
from gradio_tools import StableDiffusionPromptGeneratorTool

gradio_tool = StableDiffusionPromptGeneratorTool()
```

> We pass that instance to the `Tool.from_gradio` method:

Reviewer 2: ```suggestion
Pass the tool to the `Tool.from_gradio` method:
```

```python
from transformers import Tool

tool = Tool.from_gradio(gradio_tool)
```

> Now we can manage it exactly as we would a usual custom tool. We leverage it to improve our prompt
> ` a rabbit wearing a space suit`:

Reviewer 2: ```suggestion
Now you can use it just like any other tool. For example, let's improve the prompt  `a rabbit wearing a space suit`.
```

```python
from transformers import CodeAgent

agent = CodeAgent(llm_engine, tools=[tool], add_base_tools=True)

agent.run(
    "Improve this prompt: 'A rabbit wearing a space suit', then generate an image of it.",
)
```

The model adequately leverages the tool:
```text
==Explanation from the agent==
I will use the following  tools: `StableDiffusionPromptGenerator` to improve the prompt, then `image_generator` to generate an image according to the improved prompt.

==Code generated by the agent==
improved_prompt = StableDiffusionPromptGenerator('A rabbit wearing a space suit')
print(f"The improved prompt is {improved_prompt}.")
image = image_generator(improved_prompt)
```

Before finally generating the image:

<img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/rabbit.png">

> <Tip warning={true}>
> 
> gradio-tools requires *textual* inputs and outputs, even when working with different modalities. This implementation
> works with image and audio objects. The two are currently incompatible, but will rapidly become compatible as we
> work to improve the support.
> 
> </Tip>

Reviewer 2: ```suggestion
> [!WARNING]
> gradio-tools require *textual* inputs and outputs even when working with different modalities like image and audio objects. Image and audio inputs and outputs are currently incompatible.
```

> ### Import tools from LangChain

Reviewer 2: ```suggestion
### LangChain tools
```

> You can also use method `from_langchain`to quickly initialize a tool from a LangChain tool.
> This goes as follows: 

Reviewer 2: ```suggestion
To import a tool from LangChain, use the `from_langchain()` method.
```

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.

⚠️ Note that this file is in Markdown but contain specific syntax for our doc-builder (similar to MDX) that may not be
rendered properly in your Markdown viewer.

-->

> # Agents

Author: You're right, let's add the mention of tools! 😃

Large Language Models (LLMs) trained to perform [causal language modeling](https://huggingface.co/docs/transformers/tasks/language_modeling) can tackle a wide range of tasks, but they often struggle with basic tasks like logic, calculation, and search. The worst scenario is when they perform poorly in a domain, such as math, yet still attempt to handle all the calculations themselves.

One approach to overcome this weakness is to embed the LLM into a system where it has the ability to call tools: such a system is called an LLM Agent.

> The augmentation with tools gives the LLM superpowers. See for yourselves:

Reviewer: Nice intro

- a toolbox from which the agent pick tools to execute
- a parser to extract from the LLM output which tools are to call and with which arguments

To start with, please install the `agents` extras in order to install all default dependencies.

```bash
pip install transformers[agents]
```

To build your LLM engine, you have to define a `llm_engine` method, that will be given text and return text. This callable needs to accept a `stop` argument defining stop sequences indicating when to stop generating its output. For instance as follows:

```python
from huggingface_hub import login, InferenceClient

login("<YOUR_HUGGINGFACEHUB_API_TOKEN>")

client = InferenceClient(model="mistralai/Mixtral-8x7B-Instruct-v0.1")

def llm_engine(query: str, stop=["Task"]) -> str:
    response = client.text_generation(query, stop_sequences=stop, return_full_text=False, max_new_tokens=1000)
    for stop_seq in stop:
        if response[-len(stop_seq) :] == stop_seq:
            response = response[: -len(stop_seq)]
    return response
```

You could use any other `llm_engine` method, as long as:

- it takes a `str` as input and returns a `str`

> - it will stop generating output _before_ the sequences passed in argument `stop`

Author: Thank you. I update this to reflect the fact that we now use Messages (so a List[Dict[str, str]]) as input to the llm_engine rather than plain text.

Then you can define your agent and run it.

```python
from transformers import CodeAgent

agent = CodeAgent(llm_engine=llm_engine, tools=[], add_base_tools=True)

agent.run("Draw me a picture of rivers and lakes")
```

The prompt and output parser were automatically defined here, but you can easily inspect them.

![Untitled](Doc%20agents%20ac753b9a3b934eaba22f659ba994c4bd/Untitled%201.png)

```python
agent.prompt_template
```

Note that your agent is powered by a LLM, so small variations in your prompt might yield completely different results. It's important to explain as clearly as possible the task you want to perform. We go more in-depth on how to write good prompts [here](custom_tools#writing-good-user-inputs).

Every [`~Agent.run`] operation is independent, so you can run it several times in a row with different tasks.

## Implementation of agents

### ReactAgent

This is the go-to agent to solve reasoning tasks, since the ReAct framework makes it really efficient to think on the basis of its previous observations.

### CodeAgent

> This class is the one to use when trying to execute multimodal tasks, since it natively handles different input and output types for its tools. It has a planning step, then generates python code to execute all actions at once.

Author: Thank you! I moved that section up. Since writing the description above I've added another PR that separates ReactAgent into ReactJSONAgent and ReactCodeAgent, so I kept the structure with titles (i.e. separate `### React agents` from `### Code agent`)

> #### Code execution?!

Author: Thank you, that's changed!

This code is then executed with our small Python interpreter on the set of inputs passed along with your tools. We hear you screaming "Arbitrary code execution!" in the back, but let us explain why that is not the case.

The only functions that can be called are the tools you provided and the print function, so you're already limited in what can be executed. You should be safe if it's limited to Hugging Face tools. 

Then, we don't allow any attribute lookup or imports (which shouldn't be needed anyway for passing along  inputs/outputs to a small set of functions) so all the most obvious attacks (and you'd need to prompt the LLM  to output them anyway) shouldn't be an issue.

> The execution will stop at any line trying to perform an illegal operation or if there is a regular Python error with the code generated by the agent.

Author: Thanks a lot! I replaced the text with your suggestion.

### Default toolbox

> We curated a set of tools that can empower agents. Here is an updated list of the tools we have integrated in `transformers`:

Author: Thank you, this is integrated.

Action:
$ACTION_JSON_BLOB
Observation: the result of the action
... (this Thought/Action/Observation can repeat N times, you should take several steps when needed. The $ACTION_JSON_BLOB must only use a SINGLE action at a time.)

ALWAYS provide a 'Thought:' and an 'Action:' part. You MUST provide at least the 'Action:' part to move forward.
To provide the final answer to the task, use an action blob with "action": 'final_answer' tool.It is the only way to complete the task, else you will be stuck on a loop. So your final output should look like this:
{
  "action": 'final_answer',
  "action_input": "insert your final answer here"
}

Now begin!
```

The prompt has these parts:

- Introduction: how the agent should behave, explanation of the concept of tools.
- Description of all the tools. This is defined by a `<<tool_descriptions>>` token that is dynamically replaced at runtime with the tools defined/chosen by the user.
    - The description is built based on the tool attributes `name`, `description`, `inputs` and `output_type`  with a simple `jinja2` template, that you can refine.
- Clarifications on the output format

> - Introduction of the task at hand.

Author: Thank you, reads much better! 😃

```

<Tip warning={true}>

Please make sure to have the `<<all_tools>>` string defined somewhere in the `template` so that the agent can be aware 
of the tools, it has available to it.

</Tip>

## Adding new tools

In this section, we show how to create a new tool that can be added to the agent.

### Creating a new tool and sharing it to the Hub

We'll first start by creating a tool. We'll add the not-so-useful yet fun task of fetching the model on the Hugging Face
Hub with the most downloads for a given task.

We can do that with the following code:

```python
from huggingface_hub import list_models

task = "text-classification"

model = next(iter(list_models(filter=task, sort="downloads", direction=-1)))
print(model.id)
```

For the task `text-classification`, this returns `'facebook/bart-large-mnli'`, for `translation` it returns `'google-t5/t5-base`.

How do we convert this to a tool that the agent can leverage? All tools depend on the superclass `Tool` that holds the
main attributes necessary. We'll create a class that inherits from it:

```python
from transformers import Tool

class HFModelDownloadsTool(Tool):
    pass

> ```

Author: Thank you, indeed these lines were not necessary.

> This class has a few needs:
> - An attribute `name`, which corresponds to the name of the tool itself. To be in tune with other tools which have a
>   performative name, we'll name it `model_download_counter`.
> - An attribute `description`, which will be used to populate the prompt of the agent.
> - `inputs` attribute. This is a dictionnary with keys `"type"`and `"description"`containing information to help the python interpreter make educated choices about the input,
>   and will allow for a gradio-demo to be spawned when we push our tool to the Hub.
> - An attribute `output_type` specifying the type of the ouput. 
> - A `__call__` method which contains the inference code. This is the code we've played with above!
> 
> Here's what our class looks like now:

Reviewer 2: ```suggestion
The custom tool needs:

- An attribute `name`, which corresponds to the name of the tool itself. The name usually describes what the tool does. Since the code returns the model with the most downloads for a task, let's name is `model_download_counter`.
- An attribute `description` is used to populate the agent's system prompt.
- An `inputs` attribute, which is a dictionary with keys `"type"` and `"description"`. It contains information that helps the Python interpreter make educated choices about the input.
- An `output_type` attribute, which specifies the output type.
- A `__call__` method which contains the inference code to be executed.
```

```python
from transformers import Tool
from huggingface_hub import list_models

class HFModelDownloadsTool(Tool):
    name = "model_download_counter"
    description = (
        "This is a tool that returns the most downloaded model of a given task on the Hugging Face Hub. "
        "It returns the name of the checkpoint."
    )

    inputs = {
        "task": {
            "type": str,
            "description": "the task category (such as text-classification, depth-estimation, etc)",
        }
    }
    output_type = str

    def __call__(self, task: str):
        model = next(iter(list_models(filter=task, sort="downloads", direction=-1)))
        return model.id
```

> We now have our tool handy. Save it in a file and import it from your main script. Let's name this file
> `model_downloads.py`, so the resulting import code looks like this:

Reviewer 2: ```suggestion
Now that the custom `HfModelDownloadsTool` class is ready, you can save it to a file named model_downloads.py and import it for use.
```

```python
from model_downloads import HFModelDownloadsTool

tool = HFModelDownloadsTool()
```

> In order to let others benefit from it and for simpler initialization, we recommend pushing it to the Hub under your 
> namespace. To do so, just call `push_to_hub` on the `tool` variable (be sure to have created the space beforehand, and to be authenticated with a token that has read access!)
> 
> ```python
> tool.push_to_hub("{your_username}/hf-model-downloads")
> ```
> 
> You now have your code on the Hub! Let's take a look at the final step, which is to have the agent use it.

Reviewer 2: ```suggestion
You can also share your custom tool to the Hub by calling [`~Tool.push_to_hub` on the tool. Make sure you've created a repository for it on the Hub and have a token with read access.

```python
tool.push_to_hub("{your_username}/hf-model-downloads")
```

```

> ### Initializing an agent that uses the tool
> 
> We now have our tool that lives on the Hub which can be instantiated as such (change the user name for your tool):

Reviewer 2: ```suggestion
Load the tool with the [`~Tool.load_tool] function and pass it to the `tools` parameter in your agent.
```

> ```python
> from transformers import load_tool
> 
> model_download_tool = load_tool("m-ric/hf-model-downloads")
> ```
> 
> In order to use it in the agent, simply pass it to the agent initialization method.
> 
> ```python
> from transformers import CodeAgent
> 
> agent = CodeAgent(llm_engine, tools=[model_download_tool])
> 
> agent.run(
>     "Can you give me the name of the model that has the most downloads in the 'text-to-video' task on the Hugging Face Hub?"
> )
> ```

Reviewer 2: ```suggestion
```python
from transformers import load_tool, CodeAgent

model_download_tool = load_tool("m-ric/hf-model-downloads")
agent = CodeAgent(llm_engine, tools=[model_download_tool])
agent.run(
    "Can you give me the name of the model that has the most downloads in the 'text-to-video' task on the Hugging Face Hub?"
)
```
```

You get the following:
```text
==Executing the code below:==
task = 'text-to-video'
model_name = model_download_counter(task=task)
print(f"The most downloaded model in the '{task}' task is '{model_name}'.")
```

And the output:
`"The most downloaded model in the 'text-to-video' task is 'ByteDance/AnimateDiff-Lightning'."`

> <Tip>
> 
> Depending on the LLM, some are quite brittle and require very exact prompts in order to work well. Having a well-defined
> name and description of the tool is paramount to having it be leveraged by the agent.
> 
> </Tip>

Reviewer 2: I think it makes more sense to add this tip to when you describe what attributes the custom tool needs

```suggestion
> [!TIP]
> Some LLMs can be brittle and require very exact prompts in order to work well. It is very important that your tool has a well-defined name and description in order for the agent to successfully leverage it.
```

Author: I agree! I moved that up below the description of custom tool attributes.

> ### Adding a new tool to an existing agent's toolbox

Reviewer 2: Combine this section and the next one into a `### Manage agent toolbox` section. I'll finish reviewing the rest of the doc tomorrow :)
